diagnostics-app/src/app/core/services/events.service.ts

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Service for starting to capture events and stopping
+ * capturing events from Chrome extension.
+ */
+
+import { Injectable } from '@angular/core';
+import { Subject } from 'rxjs';
+
+import { environment } from 'src/environments/environment';
+import {
+  EventCategory,
+  EventsInfo,
+} from '@common/dpsl'
+import {
+  EventsAction,
+  EventMessage,
+  EventsRequest,
+  EventsResponse,
+  Request,
+  RequestType,
+  Response,
+  PortName,
+} from '@common/message';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class EventsService {
+  private extensionId!: string;
+  private subjects = new Map<string, Subject<EventsInfo>>;
+  private port?;
+
+  private constructEventsRequest: (
+    payload: EventsRequest
+  ) => Request = payload => {
+    return { type: RequestType.EVENTS, events: payload };
+  };
+
+  private sendRequest: (request: Request) => Promise<EventsResponse> = (
+    request: Request
+  ) => {
+    return new Promise((resolve, reject) => {
+      try {
+        //@ts-ignore
+        window.chrome.runtime.sendMessage(
+          this.extensionId,
+          request,
+          (response: Response) => {
+            if (!response.success) {
+              reject(response.error);
+            } else if (!response.events) {
+              throw 'Invalid response';
+            } else {
+              resolve(response.events);
+            }
+          }
+        );
+      } catch (err) {
+        reject(err);
+      }
+    });
+  };
+
+  constructor() {
+    this.extensionId = environment.extensionId;
+
+    try {
+      //@ts-ignore
+      this.port = window.chrome.runtime.connect(
+        environment.extensionId,
+        { name: PortName.EVENTS_PORT }
+      );
+
+      this.port.onMessage.addListener((msg: EventMessage) => {
+        if (this.subjects.has(msg.type)) {
+          this.subjects.get(msg.type)!.next(msg.info);
+        } else {
+          console.error('Error when mapping events type to get subject.');
```

Comment:
Can we define another string in `message.ts` for this kind of error?

Reply:
Done.